lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,344 @@
+    :name user})
+
+(def ^:private clojure-core-ns-usage
+  ;; add :filename and :from ns
+  '{:bucket :namespace-usages
+    :name clojure.core})
+
+(def ^:private cljs-core-ns-usage
+  ;; add :filename and :from ns
+  '{:bucket :namespace-usages
+    :name cljs.core})
+
+(defn ^:private ns-definitions-and-usages [analysis]
+  (letfn [(result-with-element [result element]
+            (let [result (if (= 'user (:ns element))
+                           (assoc result :in-user true)
+                           result)]
+              (case (:bucket element)
+                :namespace-definitions (update result :defs conj element)
+                :namespace-usages      (update result :usages conj element)
+                :var-usages            (case (:to element)
+                                         clojure.core (update result :to-clj conj (:from element))
+                                         cljs.core    (update result :to-cljs conj (:from element))
+                                         result)
+                result)))]
+    (let [{:keys [defs usages]}
+          (reduce-kv (fn [result filename elements]
+                       (let [result (reduce result-with-element
+                                            (assoc result
+                                                   :in-user false
+                                                   :to-clj #{}
+                                                   :to-cljs #{})
+                                            elements)
+                             implicit-elements (concat
+                                                 (when (:in-user result)
+                                                   [(assoc user-ns-def :filename filename)])
+                                                 (map (fn [from-ns]
+                                                        (assoc clojure-core-ns-usage
+                                                               :from from-ns
+                                                               :filename filename))
+                                                      (:to-clj result))
+                                                 (map (fn [from-ns]
+                                                        (assoc cljs-core-ns-usage
+                                                               :from from-ns
+                                                               :filename filename))
+                                                      (:to-cljs result)))]
+                         (reduce result-with-element result implicit-elements)))
+                     {:defs [] :usages []}
+                     analysis)]
+      [defs usages])))
+
+(defn refresh-analysis [db old-analysis new-analysis internal?]
+  ;; TODO: should this be wrapped in use-dep-graph? If it is, startup will be
+  ;; faster (500ms in medium projects 1200ms in v. large). But if it isn't, you
+  ;; can toggle the setting live.
```

Comment:
We agreed on keep it for consistency, so we can remove the TODO and maybe just leave as a NOTE?

Reply:
Oh, right... thanks for the reminder.